test/jasmine/assets/check_texttemplate.js

```diff
@@ -27,17 +31,30 @@ module.exports = function checkTextTemplate(mock, selector, tests) {
     tests.push(['%{customdata}', customdata]);
 
     tests.forEach(function(test) {
-        it('should support texttemplate', function(done) {
+        it(isGL ? '@gl' : '' + 'should support texttemplate', function(done) {
```

Comment:
Could you split the non-gl from the gl test in two separate `it`s (see [how `select_test.js` does it](https://github.com/plotly/plotly.js/blob/ffa6e3468b592e45db2a00bb65805a92fd59c4fb/test/jasmine/tests/select_test.js#L619-L664) . Combining them here will throw off test [sharding](https://github.com/plotly/plotly.js/blob/master/tasks/shard_jasmine_tests.js)

Reply:
@etpinard can you have a look at commit https://github.com/plotly/plotly.js/commit/05f472229f99e0f6bd82200f6192a7ec2fb17638 ? Are the non-gl tests corectly split from the gl?

Reply:
Hmm. Come to think of it, we might have to refactor `checkTextTemplate` wrapper so that each `it` is in the `_test.js` files, not in the `assets/` folder to make test sharding work as expected.

Would it be too much to ask to have (e.g. for the `sunburst_test.js` suite):

```js
describe('Test sunburst texttemplate:', function() {
  var trace = {
        type: 'sunburst',
        labels: ['Eve', 'Cain', 'Seth', 'Enos', 'Esther'],
        values: [11, 12, 13, 14, 15],
        text: ['1', '2', '3', '4', '5'],
        parents: ['', 'Eve', 'Eve', 'Seth', 'Seth' ]
   }

   var selector = 'g.slicetext'
   
   [
      ['txt: %{label}', ['txt: Eve', 'txt: Cain', 'txt: Seth', 'txt: Enos', 'txt: Esther']],
      [['txt: %{label}', '%{text}', 'value: %{value}'], ['txt: Eve', '2', 'value: 13', '', '']],
      ['%{color}', ['rgba(0,0,0,0)', '#1f77b4', '#ff7f0e', '#1f77b4', '#1f77b4']]
    ].forEach(function(test, i) {
      it('should support texttemplate - case ' + i, function(done) {
          checkTextTemplate([trace], selector, test).catch(failTest).then(done);
      });
    });
});
```

That way all the `it`s would remain in `test/jasmine/tests/` and sharding would work as expected.


You could combine the "should not coerce textinfo when texttemplate" `it` inside `checkTextTemplate` or have `assets/check_texttemplate.js` export two methods e.g. `checkTextTemplateDefaults` and `checkTextTemplateRendering`

Reply:
Ok, I now understand how `shard_jasmine_tests.js` proceeds. Unfortunately, it misses out on `it` blocks that are within required files or in `for` loops. The latter is a common pattern used throughout our tests: https://github.com/plotly/plotly.js/blob/ffa6e3468b592e45db2a00bb65805a92fd59c4fb/test/jasmine/tests/scatterpolargl_test.js#L276-L283

I think we should find a more robust way of counting the number of `it` blocks in test files... :thinking:

Reply:
> Unfortunately, it misses out on `it` blocks that are within required files or in `for` loops.

Does it really _miss_ out? Or does it just count them as one `it` test?

Reply:
> I think we should find a more robust way of counting the number of `it` blocks in test files...

I agree. The best way to do this would be to write a [karma plugin](http://karma-runner.github.io/4.0/dev/plugins.html) for this where we would have access to all the "generated" `it`s at runtime.

Reply:
@etpinard upon a second look, I'm not even sure it's worth counting the `it` at all considering most of the time there's only one file per shard (see below). Therefore, one could argue that all we need is to filter out files without any tests matching a given tag in order to skip the wasteful build/run step.

### Breakdown of shards
The folllowing line:
https://github.com/plotly/plotly.js/blob/c91f88b0f6f41bdfda46ecf28be06bf534d993c0/.circleci/test.sh#L56
yields:
```
contourgl_test.js
gl2d_click_test.js
gl2d_plot_interact_test.js
gl3d_hover_click_test.js
gl3d_plot_interact_test.js
mapbox_test.js
mesh3d_test.js
parcoords_test.js
plot_api_react_test.js
scatter3d_test.js
scattergl_select_test.js
scattergl_test.js
scattermapbox_test.js
splom_test.js
streamtube_test.js
surface_test.js
cone_test.js
scatterpolargl_test.js
isosurface_test.js,cartesian_test.js
volume_test.js,colorscale_test.js
select_test.js,gl3dlayout_test.js
pointcloud_test.js,modebar_test.js
choroplethmapbox_test.js,densitymapbox_test.js
```
Except for the last 5 lines, each shard only has one file.

The following line: https://github.com/plotly/plotly.js/blob/c91f88b0f6f41bdfda46ecf28be06bf534d993c0/.circleci/test.sh#L67
yields:
```
config_test.js
select_test.js
shapes_test.js
animate_test.js
polar_test.js
```
Then again, each shard only has one file.

Reply:
> considering most of the time there's only one file per shard (see below).

_currently_ there's only one file per shard, but there's nothing guaranteeing that we won't change `--limit` value in the future.

Oh well, I think having all the `it`s in `test/jasmine/tests/` is a good idea regardless, but I guess we can leave that for another PR if you prefer.

Reply:
I understand that we may want to increase that limit.

> I guess we can leave that for another PR if you prefer.

Cool, I think I would prefer to tackle this in another PR considering there are other files under `test/jasmine/assets/` that are not related to this PR and contain `it` specs.